### PR TITLE
Fix trace backward nonFix wrong exception types in validation codeFix for #171704: torch.trace backward produces incorrect gradients for non-square inputs square matrices

### LINKED ISSUE #171704: `torch.trace` backward produces incorrect gradients for non-square inputs.
### 🐛 Describe the bug

The backward implementation for `trace` in `aten/src/ATen/native/TriangularOps.cpp` (`trace_backward_symint`) computes linear indices for the main diagonal using a fixed step of `m + 1` (where `m = sizes[1]` and `n = sizes[0]`). For non-square inputs with `n > m`, this produces an extra index that wraps into the next row (e.g., `(3, 0)` for a `4x2` input), causing gradients to be written outside the diagonal. This yields incorrect gradients.

### Affected code
- File: `aten/src/ATen/native/TriangularOps.cpp`
- Function: `trace_backward_symint`

Relevant lines as of this report:
```cpp
auto grad_input = at::zeros_symint(sizes[0] * sizes[1], grad.options());
auto indices = at::arange(0, grad_input.numel(), sizes[1] + 1, grad.options().dtype(at::kLong));
// ...
grad_input.index_fill_(0, indices, grad);
```

### Reproduction
Minimal Python repro:

```python
import torch

x = torch.zeros(4, 2, requires_grad=True)
out = torch.trace(x)  # uses main diagonal, length = min(4, 2) = 2
out.backward()
print(x.grad)
```

#### Expected gradient
On the main diagonal only, up to `min(n, m)`:

```text
tensor([[1., 0.],
        [0., 1.],
        [0., 0.],
        [0., 0.]])
```

#### Actual gradient (current implementation)
An extra 1 appears at `(3, 0)` due to the linear index sequence `[0, 3, 6]` for `n=4, m=2, step=m+1=3`:

```text
tensor([[1., 0.],
        [0., 1.],
        [0., 0.],
        [1., 0.]])
```

#### Cross-framework comparison (JAX)
JAX produces the expected gradient for the same non-square input:

```python
import jax
import jax.numpy as jnp

x = jnp.zeros((4, 2))
g = jax.grad(jnp.trace)(x)
print(g)
```

Expected output:

```text
[[1., 0.],
 [0., 1.],
 [0., 0.],
 [0., 0.]]
```

### Versions

Observed on PyTorch 2.7.1, but I believe it still exists in the main branch.

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93

### PR DESCRIPTION
**Pull request by gaurav-redhat**
**Fix trace backward for non-square matrices**

Fixes https://github.com/pytorch/pytorch/issues/171704

Bug: For (n, m) matrices where n > m, trace_backward_symint wrote gradients outside the diagonal (e.g., position (3, 0) for a 4x2 matrix).
Fix: Use min(n, m) to compute diagonal length instead of numel().
Test: Added test_trace_backward covering shapes (4,2), (2,4), (1,5), (5,1), (3,3), (1,1).

### CODE REVIEW

**Review by janeyx99 — APPROVED** [review]
thanks!

File: test/test_shape_ops.py [hunk 1]
```diff
@@ -294,7 +294,9 @@ def test(shape):
             test(shape)
 
     @onlyNativeDeviceTypes
-    @dtypes(torch.float, torch.double)
+    @dtypes(torch.float)
```
Comment (janeyx99) [inline test/test_shape_ops.py]: how come double doesn't work on MPS? is it a known issue? If so let's link the issue?

Comment (gaurav-redhat) [reply test/test_shape_ops.py]: Fixed. Now using @dtypes(torch.float, torch.double) as default with @dtypesIfMPS(torch.float) override. Added comment linking to #115350.
